src/Runtime/Distributions/Automata/Automaton.EnumerateSupport.cs

```diff
@@ -0,0 +1,553 @@
+            // - Is this state a dead end. I.e. end state is not reachable through any path starting
+            //   from this state. This flag is computed incrementally most of the time. But if
+            //   automaton contains loops, a special procedure `ComputeEndStatesReachability` is
+            //   invoked which computes it efficiently for whole graph.
+            // - Whether this state is being visited now and if yes - how long is the sequence from
+            //   root to this state. This is used for (non-empty) loop detection. We store length
+            var flags = new StateEnumerationFlags[this.States.Count];
+
+            // Number of sequences produced by this time. By comparing value of this counter
+            // before entering the state and when leaving it is easy to detect whether the state
+            // is a dead ent.
+            var producedCount = 0;
+
+            // Enumeration state for current state. A top of traversal stack, materialized in local
+            // variable for convenience.
+            var current = InitEnumerationState(this.Start.Index, 0, this.Start.CanEnd ? 0 : -1);
+
+            // If true, we can assume that `IsDeadEnd` flag is computed for all states
+            var endStateReachabilityComputed = false;
+
+            // Mark the first state as visited
+            flags[this.Start.Index] = (StateEnumerationFlags)1;
+
+            if (this.Start.CanEnd)
+            {
+                // We do not ever "enter" the start state via a transition, so need to handle
+                // "start state is also an end state" case explicitly
+                yield return SequenceManipulator.ToSequence(path);
+            }
+
+            while (true)
+            {
+                if (!TryBacktrackIfNeed())
+                {
+                    // Nowhere to backtrack, enumerated everything
+                    yield break;
+                }
+
+                // In theory, this operation could be called only when `current.PathLength`
+                // decreases. But that happens in multiple places, so it's easier to truncate path
+                // on each iteration, even if it is a noop.
+                path.RemoveRange(current.PathLength, path.Count - current.PathLength);
+
+                if (!AdvanceToNextTransition(out var nextStateIndex, out var nextElement))
+                {
+                    // Failed to go to next transition: destination state is unreachable
+                    continue;
+                }
+
+                if (nextElement.HasValue)
+                {
+                    path.Add(nextElement.Value);
+                }
+
+                nextStateIndex = TraverseTrivialStates(nextStateIndex);
+                var nextState = this.Data.States[nextStateIndex];
+
+                if (flags[nextStateIndex] != 0)
+                {
+                    // We have found a loop, so let's check if it can produce anything.
```

Comment:
We would also step here if the nextState was previously determined to be a dead end. The code that follows correctly decides to not traverse this path further, but the comment is a bit misleading.

Reply:
Fixed comment